### PR TITLE
AMPI: replace custom complex classes with std::complex

### PR DESCRIPTION
**Pull request by matthiasdiener**

### CODE REVIEW

File: src/libs/ck-libs/ampi/ampi.C [hunk 1]
```diff
@@ -193,72 +193,6 @@ static mpi_comm_worlds mpi_worlds;
...
-    CkAbort("AMPI> Cannot compare long double complex numbers with MPI_MAX\n");
-    return 0;
-  }
-  int operator<(const AmpiLongDoubleComplex &a) noexcept {
-    CkAbort("AMPI> Cannot compare long double complex numbers with MPI_MIN\n");
-    return 0;
-  }
-};
-
 typedef struct { float val; int idx; } FloatInt;
```
Comment (stwhite91) [inline src/libs/ck-libs/ampi/ampi.C]: Could remove these?

Comment (matthiasdiener) [reply src/libs/ck-libs/ampi/ampi.C]: but aren't these struct still necessary?

Comment (stwhite91) [reply src/libs/ck-libs/ampi/ampi.C]: Ah yeah nvm I misread

**Comment by matthiasdiener** [comment]
> ddt.h also has definitions of these complex types that should match the ones here. I think we should switch to std::complex wherever possible, assuming it is compatible with MPI's definition of complex types for the C and Fortran bindings

hmm, are you sure they need to match? currently, there seem to be different definitions already (class with methods vs. struct with just fields).

**Comment by stwhite91** [comment]
> hmm, are you sure they need to match? currently, there seem to be different definitions already (class with methods vs. struct with just fields).

Only the memory layout of the structs/classes need to match with each other (ampi.C and ddt.h), and with the MPI standard. I see now that for the first problem, C++11 does guarantee that `std::complex<T>` has the same memory layout as a struct with two `T` members, so I suppose it's not really an issue though it seems nicer to have them match. I also see now that the MPI standard has language specific bindings to address the second problem for pre-C++11 and for Fortran: MPI_C_FLOAT_COMPLEX, MPI_CXX_FLOAT_COMPLEX, MPI_COMPLEX8, MPI_TYPE_CREATE_F90_COMPLEX, etc.

**Comment by matthiasdiener** [comment]
> > hmm, are you sure they need to match? currently, there seem to be different definitions already (class with methods vs. struct with just fields).
> 
> Only the memory layout of the structs/classes need to match with each other (ampi.C and ddt.h), and with the MPI standard. I see now that for the first problem, C++11 does guarantee that `std::complex<T>` has the same memory layout as a struct with two `T` members, so I suppose it's not really an issue though it seems nicer to have them match. I also see now that the MPI standard has language specific bindings to address the second problem for pre-C++11 and for Fortran: MPI_C_FLOAT_COMPLEX, MPI_CXX_FLOAT_COMPLEX, MPI_COMPLEX8, MPI_TYPE_CREATE_F90_COMPLEX, etc.

Ok, considering that the complex types in ddt are only used to calculate offsets within the structure, I think it's safer to leave them as they are (as opposed to determining the offsets from std::complex).

**Comment by matthiasdiener** [comment]
This needs another review.

**Pull request merged**